### ISSUE TITLE
Dask-mesos tool

### ISSUE THREAD

**Issue by mrocklin** [dask/distributed#320]
@hussainsultan and I were playing with https://github.com/hussainsultan/mesos-distributed .  We'd like to make it a more thoroughly tested and deployable tool.  There are a few things that need to be done:
1.  Someone (probably some combination of @hussainsultan and myself) need to walk through the code to improve handling of how it handles offers
2.  Someone (probably @hussainsultan or @danielfrg) needs to construct a nice docker system for testing that can be used both on a CI service and locally.  @hussainsultan found this [docker-compose yaml file](https://gist.github.com/e3dcfd221fadf49a3b979408d47def3c) which seems to work well on my machine and may be of some use.   The [Satyr project](https://github.com/lensacom/satyr) may also have some nice resources.
3.  We probably want a simple CLI for this as well as a pretty simple programmatic interface
4.  Eventually (not part of this issue) we'll probably want a way to convey to the system what kinds of resources we want with more precision than the number of cores.

I'm considering a renaming of all `dfoo` tools to `dask-foo`.  So if we make this a separate package (which I support) then I suggest the name `dask-mesos`, although counter-arguments are welcome as always.

**Comment by mrocklin** [comment]
This is a more specific version of https://github.com/dask/distributed/issues/66

**Comment by kszucs** [comment]
I can help You with this issue. A baseline implementation is pretty straightforward with Satyr - with some nice resources indeed. BTW there is already a [mesos backend](https://github.com/lensacom/dask.mesos) for dask, but schedules tasks instead of workers. 
For integration with distributed something like http://spark.apache.org/docs/latest/job-scheduling.html#dynamic-resource-allocation would be nice.

**Comment by mrocklin** [comment]
Help would be appreciated.

One of the biggest hurdles to getting the dask-mesos work going, I think, is a standard test environment.  When building HDFS support we ended up building a Docker setup that could be run both locally and on travis.ci (see https://github.com/dask/distributed/tree/master/continuous_integration).  This approach ended up working fairly well.  What do you all do to test and provide continuous integration for Satyr?  What do you think is the right approach here?

Do you think it is best to build the dask-mesos solution on top of Satyr or on top of the raw Python bindings?  What are the costs and benefits of each?

I agree that dynamic resource allocation would be good to have.  I would like for Dask to handle this problem separately from Mesos.  There are several systems that would benefit from this logic.  I'm waiting to develop dynamic resource allocation until we have a couple resource managers like Mesos set up.

**Comment by kszucs** [comment]
Drone supports local execution too (it only needs a docker daemon then `drone exec --trusted`). We had a great experience with it https://github.com/lensacom/satyr/blob/master/.drone.yml - syntax is similar to docker-compose.  We use a self-hosted version because dind requires priviled (trusted) mode. Is there a chance to use anything else than travis?

Satyr is build upon `mesos.native`, the raw Python binding (which has changed since the release of mesos-1.0.0). The interface of mesos native is catastrophic(!) (thanks to C++ syntax and protobuf), but Satyr wraps it with a more pythonic one.

```py
from distributed import Scheduler as DistributedScheduler
from distributed import Worker as DistributedWorker

from satyr.proxies.messages import Disk, Mem, Cpus
from satyr.scheduler import QueueScheduler, Running
from satyr.messages import PythonTask


framework = QueueScheduler()

with Running(framework, name='dask-distributed'):
    ds = DistributedScheduler(loop=loop)
    ds.start(port)  # does it block?

    for i in num_workers:
        # it could be a nanny process istead of DistributeWorker
        task = PythonTask(name='dworker', fn=DistributedWorker, args=[host, port],
                          resources=[Cpus(1), Mem(512), Disk(1024)],
                          docker='dask/distributed')
        framework.submit(task)

    framework.wait()  # I guess it should be the distributed's loop
```

All parts of Satyr are easily customizable. Just compare satyr's [scheduler](https://github.com/lensacom/satyr/blob/master/satyr/scheduler.py#L61) implementation (can run custom callbacks, docker executors, supports bin-packing and task retries) with the [reference](https://github.com/apache/mesos/blob/master/src/examples/python/test_framework.py#L32) one. A more realistic example of writing a mesos framework is [changes-mesos-framework](https://github.com/dropbox/changes-mesos-framework/blob/master/changes_mesos_scheduler/changes_scheduler.py). 
Additionally the event handling is in the Task itself.

**Comment by mrocklin** [comment]
Travis is nice because it's convenient to reduce the number of tools that long-term dask developers need to know.  I and several others can diagnose and fix Travis errors easily; the community of developers that know drone is smaller and so maintenance costs there become more expensive.  However, if you're willing to do the work to set up drone and include nice instructions for its use, and if you feel strongly that it's nicer to use for this than Travis then I'm happy to yield to your judgement (I suspect that you know much more about Docker and Mesos than I do.)  In the worst case scenario we can always change to some other CI service in the future.

If we use Satyr then Dask will probably impose some constraints, like Python 2/3 support.  How actively used is Satyr?  What are the odds that Satyr will still be actively developed a year from today?

The people I've asked to take a look have had positive things to say about Satyr, so I'm enthusiastic.  I just want to make sure that this is a reasonable long-term decision.

**Comment by kszucs** [comment]
Hehe :) great catch!  Python 3 is not supported, see https://issues.apache.org/jira/browse/MESOS-355
AFAIK Satyr is actively used only internally. Sadly Python is not the primary language for Mesos frameworks (mostly JVM languages). The odds are great, the bus factor is low :)

**Comment by mrocklin** [comment]
@kszucs what do you think are the first steps to build a dask-mesos manager with Satyr?  Create new repository, set up Drone CI, make a simple framework like what you propose above?

**Comment by kszucs** [comment]
Pushing Mesos community would be great :)
Drone setup is really just a copy&paste with a short guide.
First of all I need to get familiar with distributed. A minimal specification is necessary. 
 - Are there any plans to support different kind of resource managers with the same interface? 
 - What about fault tolerancy? 
 - Executor reuse?
 - Stateful components?
 - High-Availability?

Could You give me a snippet how should the interface look like? E.g.: 

```python
distributed.MesosScheduler(
  executor_cores=4, 
  executor_memory=4*GiB, 
  executor_docker_image='dask/distributed',
  cores_max=64, 
  constraints={'host': 'unique', ...},
  ...)
```

**Comment by mrocklin** [comment]
Right, so a few useful links:

1.  An old issue calling out for a uniform resource manager interface: https://github.com/dask/distributed/issues/128
2.  A newer implementation of a "LocalCluster" interface that seems to work nicely: https://github.com/dask/distributed/blob/master/distributed/deploy/local.py
3.  Current setup how-to docs: http://distributed.readthedocs.io/en/latest/setup.html

> Are there any plans to support different kind of resource managers with the same interface?

Yes, I would like to.  I think that we need to try out a few very different systems, learn from that experience, and then refactor into a common interface.

> What about fault tolerancy?

Every component can currently go down and come back up and everyone will reconnect nicely.  If the scheduler goes down it's nice if it comes back up with the same address that all of the workers used originally.

> Executor reuse?

Dask someone confusingly uses the term Executor to mean Client.  Is that what you mean here?  If so then multiple Executors can connect to the Scheduler, but this is mostly on the users to do.  I don't think that we need to care.

> Stateful components?

They're all stateful.  The scheduler has the logic to handle the loss of a worker.  If the scheduler goes down all state gets cleared.

> High-Availability?

As mentioned above.  The state in the scheduler is vulnerable.  It's nice if the scheduler comes back up in the same place where the workers expect it to be.

**Comment by mrocklin** [comment]
I think that the MesosScheduler would keep one `Scheduler` running on a machine with a single core and some memory.  

    >>> c = MesosScheduler()
    >>> c.scheduler_address
    '192.128.0.1:8786'

We would then add and remove worker requirements to it dynamically.

    >>> workers = c.start_workers(processes=10, threads_per_process=4, memory=...)
    >>> workers
    some-token

This would return immediately while the MesosScheduler tried to fulfill that request.  We might in the future ask it either to remove those workers or to remove any workers as it sees fit

    >>> c.remove_workers(workers)

I don't have a standard docker container for deployment.  My hope is that we can default to something relatively simple and then have the workers deploy software-only environments on their own in other work.  See https://github.com/dask/distributed/issues/300 . 

Thoughts?

**Comment by kszucs** [comment]
HA Scheduler is possible with mesos, see http://mesos.apache.org/documentation/latest/high-availability-framework-guide/

I meant [mesos executor](https://github.com/lensacom/satyr/blob/master/satyr/executor.py#L56) under `executor` described by [ExecutorInfo proto](https://github.com/apache/mesos/blob/0.28.x/include/mesos/mesos.proto#L419) 
The distributed worker will run inside the mesos executor (with or without nanny), telemetrics, health checks etc. If a worker shuts down in the executor the mesos scheduler can reuse the same executor to run another task. I guess this should be handled by nanny? We can also start nanny processes directly via the [CommandExecutor](https://github.com/apache/mesos/blob/0.28.x/include/mesos/mesos.proto#L353).


```>>> workers = c.start_workers(processes=10, threads_per_process=4, memory=...)```
One worker runs one process with multiple threads? 
Is it possible to run multiple workers on the same host?

For example Mesos starts an executor with 4 CPUs and 8 GiB ram. 4 processes (16 threads) and 2GiB ram per process. The 4 process will live in the same isolated environment (executor) and shared disk. On another host Mesos can start an executor with 6 CPUs and 12GiB ram and fulfills the required 10 processes with two executors. Of course we can start 4 executors on the first host and 6 on the second with allocated 1 cpu and 2GiB ram for each executor, but without shared disk. 
Which one is the preferred? 
Multiple workers inside an executor or just one? - In the first case workers can steal each others resources, e.g. memory for a shuffling stage.

We can have multiple docker containers with automated builds for different setups (conda, ubuntu-pkgs, pip) like our [pyspark](https://hub.docker.com/r/lensa/pyspark/tags/) images. If you prefer to ship code dependencies from code we can use [mesos fetcher uri](http://mesos.apache.org/documentation/latest/fetcher/) for that, it can use http, ftp, hdfs to download artifacts into the executor's sandbox. 

BTW docker is not mandatory for the executors. We can use multiple [containerizers](https://github.com/apache/mesos/blob/0.28.x/include/mesos/mesos.proto#L1620),  there is a mesos one which runs on the host and have more isolation abilities than the docker containerizer. Since mesos 1.0.0 there is a [unified containerizer](http://mesos.apache.org/documentation/latest/container-image/) too.

**Comment by mrocklin** [comment]
> The distributed worker will run inside the mesos executor (with or without nanny), telemetrics, health checks etc. If a worker shuts down in the executor the mesos scheduler can reuse the same executor to run another task. I guess this should be handled by nanny?

It would be nice to have the nannies, yes.  They provide some services that the workers can't perform themselves.  I can't think of a case when we would want to kill a nanny and then recreate another one.

> One worker runs one process with multiple threads? Is it possible to run multiple workers on the same host?

Yes, both are common.  

> Multiple workers inside an executor or just one? - In the first case workers can steal each others resources, e.g. memory for a shuffling stage.

They're both common.  Ideally this is all user-configurable.

> We can have multiple docker containers with automated builds for different setups (conda, ubuntu-pkgs, pip) like our pyspark images. If you prefer to ship code dependencies from code we can use mesos fetcher uri for that, it can use http, ftp, hdfs to download artifacts into the executor's sandbox.

Shipping around software environments is a problem that Dask will need to solve for a dozen different systems.  I'm willing to solve it internal to Dask so that we don't need to keep up with each deployment solutions' approach. 

> BTW docker is not mandatory for the executors. We can use multiple containerizers, there is a mesos one which runs on the host and have more isolation abilities than the docker containerizer. Since mesos 1.0.0 there is a unified containerizer too.

I don't have strong thoughts on how to best containerize.  I suspect that users will have opinions.  In all cases it's probably nice to pass through options to the user while also having reasonable defaults.

**Comment by kszucs** [comment]
I'm going to create a base implementation in dask.mesos with an interface like LocalCluster. User must define n_workers and the framework will group them into executors. I guess more worker in the same executor is the preferred instead of more executors with less workers, right?

Currently resizing executors is not possible, so removing a worker from an executor would change the resources per worker(task) ratio within that executor. In the first draft I won't implement remove_workers.

Is there a way to determine idle workers? To support fair scheduling it could be nice to automatically stop workers/executors.

```
Executor - mesos or docker isolated environment
Task - command run inside executor (an executor can run multiple tasks)
```
In our case a Task runs multiple workers (nannys) and an executor runs only one task (group of workers) with predefined recources for a worker.

**Comment by kszucs** [comment]
Which one is preferred strategy, launching executors with different number of workers or with fixed sized ones? 

E.g. creating 16 workers which one is better?
```py
Case 'a': Executor1(n_workers=10), Executor2(n_workers=4), Executor3(n_workers=2)
Case 'b': Executor1(n_workers=4),  Executor2(n_workers=4), Executor3(n_workers=4), Executor4(n_workers=4)

**Comment by mrocklin** [comment]
ccing @quasiben who was also building up testing infrastructure for mesos

**Comment by mrocklin** [comment]
> I guess more worker in the same executor is the preferred instead of more executors with less workers, right?

It depends.  Inter-worker bandwidth is usually better between workers on the same physical machine. 

> Is there a way to determine idle workers? To support fair scheduling it could be nice to automatically stop workers/executors.

One way that is accessible now is to ask the worker for its `.processing` dictionary, which maps workers to what tasks they're working on currently.  Empty task sets would imply idle workers.

    e = distributed.Executor(scheduler_address)
    processing = e.processing()

We can also expose more information if it would be useful

> E.g. creating 16 workers which one is better?

I don't know.  It probably depends.  I would make an arbitrary choice and then we can change it later if necessary

**Comment by quasiben** [comment]
@kszucs is there a CI service for drone or something which we can push into TravisCI?

**Comment by kszucs** [comment]
@mrocklin I'm initializing the extended LocalCluster with zero workers, but more flexible BaseCluster implementation (with loop handling, diagnostics, context management and get_ip) would be a great help. It uses a hardcoded ip, and I want to reduce the glue code.

**Comment by kszucs** [comment]
BTW good morning guys :)
@quasiben I don't understand. Drone is a complete CI solution. dask.mesos builds run in our self-hosted drone. I've seen the new dask/dask-mesos. To run locally is much like docker-comose. I can give you credentials to our drone service. To reach a broader community we should use travis, but porting would be complicated (I didn't use travis in the last year).

**Comment by kszucs** [comment]
```py
Case 'a': Executor1(n_workers=10), Executor2(n_workers=4), Executor3(n_workers=2)
Case 'b': Executor1(n_workers=4),  Executor2(n_workers=4), Executor3(n_workers=4), Executor4(n_workers=4)
```
For case 'a' I can use the current scheduler, for case 'b' I need to write another one (currently in progress).

**Comment by quasiben** [comment]
@kszucs, i meant to ask if there was a hosted solution

**Comment by mrocklin** [comment]
Is there a place I can go to see your current work?

**Comment by kszucs** [comment]
I'm planning to push later today. I'm not sure how to compose the mesos tasks. Maybe I should adapt [this code](https://github.com/dask/distributed/blob/master/distributed/cli/dask_worker.py#L49) as the mesos executor and start individual Nannys (workers) as mesos tasks. I hope this way removing a worker will be possible.

**Comment by mrocklin** [comment]
@quasiben https://drone.io/

**Comment by quasiben** [comment]
Ah!  i was trying to find that info from their docs page

**Comment by kszucs** [comment]
Right now I'm adding Mesos Unified Containerizer support to satyr. It's pretty complicated to play nicely with the build system, see http://mesos.apache.org/documentation/latest/container-image/

**Comment by kszucs** [comment]
There is a mesos issue with provisioning aufs layers of a docker image: https://issues.apache.org/jira/browse/MESOS-5028 . All docker images with whiteout files are errored, so I've managed to build an alpine image with mesos python module [kszucs/mesos-alpine:python](https://github.com/kszucs/mesos-alpine/blob/master/Dockerfile.optimized). 

Afterwards I ran into an annoying (osx + docker-machine + virtualbox) issue about not updating mounted volumes... 

A hackable  version of distributed scheduler on mesos is almost done BTW.

**Comment by mrocklin** [comment]
> There is a mesos issue with provisioning aufs layers of a docker image: https://issues.apache.org/jira/browse/MESOS-5028 . All docker images with whiteout files are errored (like the ubuntu base images from mesosphere), so I've managed to build an alpine image with mesos python module kszucs/mesos-alpine:python.

> Afterwards I ran into an annoying (osx + docker-machine + virtualbox) issue about not updating mounted volumes...

You're making me very happy that I'm not dealing with this right now :)

> A hackable version of distributed scheduler on mesos is almost done BTW.

Very exciting.  I'm looking forward to fooling around with it.

**Comment by mrocklin** [comment]
Any update @kszucs I'm playing with mesos now just using a docker-compose setup with the pure Python mesos library.  

If you have anything of interest I'd love to take a look, even if it isn't functional.

**Comment by kszucs** [comment]
@mrocklin I'm waiting on dask.mesos gitter channel https://gitter.im/lensacom/dask.mesos

**Comment by kszucs** [comment]
@mrockling First working version: https://github.com/lensacom/dask.mesos/tree/distributed/dask_mesos/distributed dinamically adding and removing workers. If You are satisfied with the implementation, we'd move this conversation to dask.mesos and plan the final interface.

**Comment by mrocklin** [comment]
Cool.  What do I need running on the Mesos master and slaves to try this
locally?

**Comment by mrocklin** [comment]
https://github.com/mrocklin/dask-marathon
